### ISSUE TITLE
statsd gauges with both arbitrary and delta values are inaccurate

### ISSUE THREAD

**Issue by RussellRollins** [vectordotdev/vector#20096]
### A note for the community

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Problem

When setting a statsd gauge to a specific value, then incrementing or decrementing it, it seems like Vector is actually storing two separate internal gauges (one for the direct set, one for the delta values).

For instance, using the example documented here:

https://github.com/statsd/statsd/blob/master/docs/metric_types.md#gauges

Instead of setting the gauge from 333 to 327, Vector actually sets the gauge to -6:

```
$ ./vector_statsd_gauges.sh 
setting gauge value to 100, expect 100
test_gauge 100
setting gauge value to 333, expect 333
test_gauge 333
incrementing gauge value by -10, expect 323
test_gauge -10
incrementing gauge value by +4, expect 327
test_gauge -6
```

Using this test script:

```
#! /usr/bin/env bash
set -euo pipefail

echo "setting gauge value to 100, expect 100"
echo "test_gauge:100|g" | nc -w 1 -u 127.0.0.1 9125
curl -sSL 127.0.0.1:9090/metrics | grep "^test_gauge"

echo "setting gauge value to 333, expect 333"
echo "test_gauge:333|g" | nc -w 1 -u 127.0.0.1 9125
curl -sSL 127.0.0.1:9090/metrics | grep "^test_gauge"

echo "incrementing gauge value by -10, expect 323"
echo "test_gauge:-10|g" | nc -w 1 -u 127.0.0.1 9125
curl -sSL 127.0.0.1:9090/metrics | grep "^test_gauge"

echo "incrementing gauge value by +4, expect 327"
echo "test_gauge:+4|g" | nc -w 1 -u 127.0.0.1 9125
curl -sSL 127.0.0.1:9090/metrics | grep "^test_gauge"
```

Started Vector via Docker as:

```
docker run -it \
    -p 9090:9090 \
    -p 9125:9125/udp  \
    -e "VECTOR_CONFIG_DIR=/etc/vector" \
    -v "$(pwd)/vector.toml:/etc/vector/vector.toml" \
    timberio/vector:0.36.1-distroless-libc
```

### Configuration

```text
# ./vector.toml

[sources.statsd]
type = "statsd"
address = "0.0.0.0:9125"
mode = "udp"

[sinks.prometheus]
type = "prometheus_exporter"
inputs = [ "statsd" ]
address = "0.0.0.0:9090"
```

### Version

vector 0.36.1 (x86_64-unknown-linux-gnu 2857180 2024-03-11 14:32:52.417737479)

**Comment by jszwedko** [comment]
Thanks @RussellRollins . I see, using the `console` sink, that the messages are being parsed correctly:

```
{"metric":{"name":"test_gauge","kind":"absolute","gauge":{"value":100.0}}}
{"metric":{"name":"test_gauge","kind":"absolute","gauge":{"value":333.0}}}
{"metric":{"name":"test_gauge","kind":"incremental","gauge":{"value":-10.0}}}
{"metric":{"name":"test_gauge","kind":"incremental","gauge":{"value":4.0}}}
```

So it seems like it is the `prometheus_exporter` sink is not aggregating them correctly.

**Comment by RussellRollins** [comment]
Hello @jszwedko, I appreciate you taking a look. I got the same results as you using the console sink, definitely possible that I am misattributing the problem in the title here and this is actually an issue with the Prometheus sink, not the statsd source.

**Comment by RussellRollins** [comment]
I'm not sure I have the skill to solve the problem, but I was able to transform the test case we've been talking about into a failing Rust test in PrometheusExporter.

I hope that might make the issue more concrete to discuss and help out someone who knows more Rust than me figure things out! I think if that test was passing, this would be fixed.

**Comment by RussellRollins** [comment]
First off @jszwedko, let me just thank y'all again for the attention on this issue, looks like there is now:

https://github.com/vectordotdev/vector/pull/20193

With a proposed change to the normalization functions.

Now that I've been friendly, if you'll permit me to be a tiny bit annoying, it's difficult for me to provide concrete examples like this. My team is in a kind of middleware-y devops-y role with the metrics stack, and we support a lot of internal customers who are using metrics for a ton of use cases. As we are helping them onboard onto a system with Vector as a key component, they have been doing some validation and edge case testing and found the issue.

However, one example I can give is from the statsd docs:

> Note:
> 
> This implies you can't explicitly set a gauge to a negative number without first setting it to zero.

So the best way to set gauges to absolute negative numbers involves a hop through absolute zero and then an incremental decrease to the goal number:

```
freezing_gaugor:0|g
freezing_gaugor:gaugor:-30|g
```

While you could just have a positive gauge for "degrees_below_zero", but maybe this is helpful as an example of a use case that is documented as using a mix of absolute and incremental values.

**Comment by jszwedko** [comment]
Thanks for the additional thoughts!

> However, one example I can give is from the statsd docs:
> 
> > Note:
> > This implies you can't explicitly set a gauge to a negative number without first setting it to zero.
> 
> So the best way to set gauges to absolute negative numbers involves a hop through absolute zero and then an incremental decrease to the goal number:
> 
> ```
> freezing_gaugor:0|g
> freezing_gaugor:gaugor:-30|g
> ```
> 
> While you could just have a positive gauge for "degrees_below_zero", but maybe this is helpful as an example of a use case that is documented as using a mix of absolute and incremental values.

I think that could also be done via:

```
freezing_gaugor:+0|g
freezing_gaugor:gaugor:-30|g
```

I would love to hear some real-world use-cases for mixing absolutes and deltas to help motivate addressing it, but we are looking into options to see if we can easily support it in https://github.com/vectordotdev/vector/pull/20099.

**Comment by RussellRollins** [comment]
I believe that would only work if the `freezing_gaugor` was already at zero. Let's say it was at -30 and you wanted to set it to an absolute value of -60.

```
freezing_gaugor:+0|g
freezing_gaugor:gaugor:-30|g
freezing_gaugor:+0|g
freezing_gaugor:gaugor:-60|g
```

You'd end up at -90. You'd have to check the current value in order to normalize it to absolute yourself by doing -30 again.

**Comment by jszwedko** [comment]
> I believe that would only work if the `freezing_gaugor` was already at zero. Let's say it was at -30 and you wanted to set it to an absolute value of -60.
> 
> ```
> freezing_gaugor:+0|g
> freezing_gaugor:gaugor:-30|g
> freezing_gaugor:+0|g
> freezing_gaugor:gaugor:-60|g
> ```
> 
> You'd end up at -90. You'd have to check the current value in order to normalize it to absolute yourself by doing -30 again.

Mmm, yeah, that's fair. This seems like a very odd rough edge of statsd to require two packets to set an absolute negative number.

**Comment by lukesteensen** [comment]
Fixed in #20193

**Comment by RussellRollins** [comment]
Wanted to chime in and say this looks good on today's 0.38.0 release, thanks again for the help!
